### PR TITLE
Optimize1

### PR DESCRIPTION
**Pull request by nathandunn**
Mostly just removed un-necessary recursions over SQL calls.  If you could take a quick look, that would be great.  Plan to retag RC3 on Wednesday. 
For #137

### CODE REVIEW

**Comment by nathandunn** [comment]
Of note:
- https://github.com/GMOD/Apollo/pull/504/files#diff-b5552b4e9fad5ee11f3c83eb66cb01fdL884
- https://github.com/GMOD/Apollo/pull/504/files#diff-2470eb92ab8820bc3971688209948114L64  (had little effect)
- https://github.com/GMOD/Apollo/pull/504/files#diff-a77ab9cd7ca3987c3800015bb38006bfL32
- https://github.com/GMOD/Apollo/pull/504/files#diff-093a3cf9e0a83a8d8bbbecc8575e51a9L44  (important for merging joins)
- https://github.com/GMOD/Apollo/pull/504/files#diff-5059381e0b30986d68f9bb0d6cdf6000R484  (removed recursive nature)

**Comment by nathandunn** [comment]
FYI @cmdcolin   @deepakunni3   Can you take a look when you have a minute.  Will merge on RC3 on Wed so today or tomorrow would be great.   Tried to highlight important sections.   

Timings are done with JMeter using the "BLAST" test based on @deepakunni3  test.  All 16 threads go as fast as they can with 1 add and 5 gets and 1 get over 1 minute.   I'll rerun the other test (10 hour 15 ops / min) tonight to make sure there aren't any weird side-effects.

**Comment by cmdcolin** [comment]
Deepak won't be back until wednesday.

Just curious to hear from you, but why do we now have a "transcript centric" processing mechanism. Are there optimization advantages to this?

**Comment by nathandunn** [comment]
The big slowdown is when recursively processing the Json from the exons CDS and no canonical splice sites.  

You could optimize for the others as well, but doing just a single join helps immensely.  

Nathan

File: grails-app/services/org/bbop/apollo/TranscriptService.groovy [hunk 1]
```diff
@@ -470,4 +474,145 @@ class TranscriptService {
...
         return (Transcript) featureRelationshipService.getParentForFeature(cds, ontologyIds as String[])
     }
+
+    /**
+     * @param gsolFeature
+     * @param includeSequence
+     * @return
+     */
+    @Timed
+    JSONObject convertTranscriptToJSON(Transcript gsolFeature, List<Feature> childFeatures = new ArrayList<>(),List<FeatureLocation> childFeatureLocations = new ArrayList<>()) {
```
Comment (nathandunn) [inline grails-app/services/org/bbop/apollo/TranscriptService.groovy]: @cmdcolin   As to your other comment, passing in the child feature locations removed a SIGNIFICANT amount of re-querying.   True, this won't work if more than one FeatureLocation, but that is kind of a global problem to solve.

**Comment by cmdcolin** [comment]
I guess with regards to this recursion issue, I felt like the recursion isn't necessarily that bad, it's just that the lazy fetching was hitting the database too many times in this case (whereas eager fetching would make it faster).  Whether recursive eager fetching is possible, i haven't fully explored, but that was just my exploration of issues on the #460

Now, with these new changes, the convertTranscriptToJSON takes a sort of hardline approach to avoid this recursion, but I think it misses subfeatures-of-subfeatures. As an example, I think the stop codon readthroughs are missed

I tested this by adding a stop codon readthrough, which works, but when I refresh, it is not displayed on the browser (when using optimize1 branch)

Now...whether we even want to enable subfeatures-of-subfeatures is another question (see https://github.com/GMOD/jbrowse/issues/584) but i think this is currently a bug on this branch

**Comment by nathandunn** [comment]
Yeah, I *JUST* fixed and pushed something related to that.  

Nathan

**Comment by cmdcolin** [comment]
Hm...stop codon is still disappearing for me on refresh. Does show up in gff3 export though.

**Comment by nathandunn** [comment]
Can you check on other versions . .  including icebox both WA2 and WA1?   I've had them appear and working, so I think (hope) it might be unrelated to optimization:

http://icebox.lbl.gov/Apollo2/annotator/index
http://icebox.lbl.gov/WebApolloDemo/annotator/index

**Comment by cmdcolin** [comment]
The issue at hand: stop codon readthrough feature does not show up i.e. the little purple icon (note that if a new stop codon readthrough is created, it does appear, but on page refresh, it is gone)

status of this issue on various servers:

- icebox WA2 works fine
- our genomes.missouri.edu with 2.0.0-RC3 snapshot works fine
- webapollodemo (WA1) works fine
- optimization1 branch on local machine does not work

**Comment by nathandunn** [comment]
Can you show me a screenshot or the location / gene and I will see if I can't fix it?  

Thanks.

**Comment by nathandunn** [comment]
has been merged for #137

**Comment by nathandunn** [comment]
Note: I manually merged this.

**Pull request closed without merging**